### PR TITLE
[ros2] grid_map_visualization port

### PR DESCRIPTION
**Pull request by Marwan99**
Changes:
- Unsubscribing from topics is not available atm in rclcpp (according to [this](https://answers.ros.org/question/354792/rclcpp-how-to-unsubscribe-from-a-topic/)), so the subscription updating functionality in `GridMapVisualization` has been disabled for now.
- The parameters handled API in ROS2 is slightly different and so the parameters have been slightly modified, example:
  From this:

    ```
    grid_map_visualizations:
      - name: elevation_points
          type: point_cloud
          params:
            layer: elevation
      - name: elevation_grid
          type: occupancy_grid
          params:
            layer: elevation
            data_min: 0.08
            data_max: -0.16
    ```

    To this:
    ```
    grid_map_visualization:
        ros__parameters:
            grid_map_topic: /grid_map_simple_demo/grid_map
            grid_map_visualizations: [elevation_points, elevation_grid]
            elevation_points:
                type: point_cloud
                params:
                    layer: elevation
            elevation_grid:
                type: occupancy_grid
                params:
                    layer: elevation
                    data_min: 0.08
                    data_max: -0.16
    ```
  Is that okay??
- Since [rclcpp::publisher](http://docs.ros2.org/foxy/api/rclcpp/classrclcpp_1_1Publisher.html) is a template, I have moved the publisher to live in the visualisation derived classes instead of having it in the `VisualizationBase` class.

### CODE REVIEW

**Comment by SteveMacenski** [comment]
> Unsubscribing from topics is not available atm in rclcpp 

You could reset the shared ptr to the subscriber to "unsubscribe" and re-create it to subscribe again.

The rest of the comments sounds good

**Comment by Marwan99** [comment]
@SteveMacenski Please review.

File: grid_map_visualization/src/GridMapVisualization.cpp [hunk 1]
```diff
@@ -6,165 +6,170 @@
...
+rclcpp::Node::SharedPtr
+GridMapVisualization::get_node_ptr()
+{
+  return nodeHandle_;
+}
+
 bool GridMapVisualization::readParameters()
 {
-  nodeHandle_.param("grid_map_topic", mapTopic_, string("/grid_map"));
+  nodeHandle_->declare_parameter("grid_map_topic", std::string("/grid_map"));
```
Comment (SteveMacenski) [inline grid_map_visualization/src/GridMapVisualization.cpp:51]: Please rename to `node`, nodehandles dont exist anymore and this won't age well

Comment (Marwan99) [reply grid_map_visualization/src/GridMapVisualization.cpp:51]: sure.

File: grid_map_visualization/src/visualizations/FlatPointCloudVisualization.cpp [hunk 2]
```diff
@@ -6,52 +6,61 @@
...
+      "did not find a 'height' parameter. Using default.",
+      name_.c_str());
   }
 
   return true;
 }
 
 bool FlatPointCloudVisualization::initialize()
 {
-  publisher_ = nodeHandle_.advertise<sensor_msgs::PointCloud2>(name_, 1, true);
```
Comment (Marwan99) [inline grid_map_visualization/src/visualizations/FlatPointCloudVisualization.cpp:40]: Just noticed the publishers were latched, do we need them latched?

Comment (SteveMacenski) [reply grid_map_visualization/src/visualizations/FlatPointCloudVisualization.cpp:40]: Good q- Max?

**Comment by Marwan99** [comment]
@maximilianwulf please review.

**Comment by maximilianwulf** [comment]
On it now.

File: grid_map_visualization/src/GridMapVisualization.cpp [hunk 3]
```diff
@@ -6,165 +6,170 @@
...
-#include "grid_map_visualization/GridMapVisualization.hpp"
 #include <grid_map_core/GridMap.hpp>
 #include <grid_map_ros/GridMapRosConverter.hpp>
 
-using namespace std;
-using namespace ros;
+#include <string>
+#include <vector>
+#include <memory>
+#include <chrono>
```
Comment (maximilianwulf) [inline grid_map_visualization/src/GridMapVisualization.cpp:15]: Alphabetical order, can you check the other includes as well?

File: grid_map_visualization/src/GridMapVisualization.cpp [hunk 4]
```diff
@@ -6,165 +6,170 @@
...
+      visualizationsParameter_.c_str(), type.c_str(), name.c_str());
   }
-
   return true;
 }
 
 bool GridMapVisualization::initialize()
 {
-  for (auto& visualization : visualizations_) {
+  for (auto visualization : visualizations_) {
```
Comment (maximilianwulf) [inline grid_map_visualization/src/GridMapVisualization.cpp:126]: Below you did not remove the `&`

File: grid_map_visualization/src/GridMapVisualization.cpp [hunk 5]
```diff
@@ -6,165 +6,170 @@
...
+    node_->get_logger(),
+    "Grid map visualization received a map (timestamp %f) for visualization.",
+    rclcpp::Time(message->header.stamp).seconds());
   grid_map::GridMap map;
-  grid_map::GridMapRosConverter::fromMessage(message, map);
+  grid_map::GridMapRosConverter::fromMessage(*message, map);
 
-  for (auto& visualization : visualizations_) {
-    visualization->visualize(map);
+  for (std::size_t i = 0; i < visualizations_.size(); i++) {
```
Comment (maximilianwulf) [inline grid_map_visualization/src/GridMapVisualization.cpp:170]: Did it not work the other way around?

File: grid_map_visualization/src/visualizations/FlatPointCloudVisualization.cpp (hunk 2 above)
Comment (maximilianwulf) [reply grid_map_visualization/src/visualizations/FlatPointCloudVisualization.cpp:40]: If possible yes, for visualization having it latched is nice.

File: grid_map_visualization/src/visualizations/FlatPointCloudVisualization.cpp [hunk 6]
```diff
@@ -6,52 +6,61 @@
...
+  publisher_ = node_->create_publisher<sensor_msgs::msg::PointCloud2>(name_, 1);
   return true;
 }
 
-bool FlatPointCloudVisualization::visualize(const grid_map::GridMap& map)
+bool FlatPointCloudVisualization::visualize(const grid_map::GridMap & map)
 {
-  if (!isActive()) return true;
```
Comment (maximilianwulf) [inline grid_map_visualization/src/visualizations/FlatPointCloudVisualization.cpp:46]: I wonder if that should actually return false.

File: grid_map_visualization/src/visualizations/FlatPointCloudVisualization.cpp [hunk 7]
```diff
@@ -6,52 +6,61 @@
...
 }
 
 FlatPointCloudVisualization::~FlatPointCloudVisualization()
 {
 }
 
-bool FlatPointCloudVisualization::readParameters(XmlRpc::XmlRpcValue& config)
+bool FlatPointCloudVisualization::readParameters()
 {
-  VisualizationBase::readParameters(config);
```
Comment (maximilianwulf) [inline grid_map_visualization/src/visualizations/FlatPointCloudVisualization.cpp:28]: Why don't we need this anymore here?

File: tools/ros2_dependencies.repos [hunk 8]
```diff
@@ -2,7 +2,7 @@ repositories:
   ros-planning/navigation2:
     type: git
     url: https://github.com/ros-planning/navigation2.git
-    version: master
+    version: main
```
Comment (maximilianwulf) [inline tools/ros2_dependencies.repos:5]: Interesting, do you know if this is this temporary or not?

**Review by maximilianwulf — COMMENTED** [review]
Done. Some comments, but nonblocking ones. Thanks a lot again for the great effort!

Please ping me via mail if you addressed the comments.

File: tools/ros2_dependencies.repos (hunk 8 above)
Comment (Marwan99) [reply tools/ros2_dependencies.repos:5]: This should be permanent, I think other packages will be making this change as well AFAIK.

File: grid_map_visualization/include/grid_map_visualization/visualizations/VisualizationBase.hpp [hunk 9]
```diff
@@ -50,67 +52,21 @@ class VisualizationBase
...
 
- protected:
-
-  /*!
-   * Get a visualization parameter as a string.
-   * @param[in] name the name of the parameter
-   * @param[out] value the string to set with the value.
-   * @return true if parameter was found, false otherwise.
-   */
-  bool getParam(const std::string& name, std::string& value);
```
[the remark this answers is not in the document]
Comment (Marwan99) [reply grid_map_visualization/include/grid_map_visualization/visualizations/VisualizationBase.hpp:69]: Yes, getting parameters is handled differently in the [ROS2 API](http://docs.ros2.org/foxy/api/rclcpp/classrclcpp_1_1Node.html#a3158e5520c58531ae4d85a2e98cdf210).

File: grid_map_visualization/src/visualizations/FlatPointCloudVisualization.cpp (hunk 6 above)
Comment (Marwan99) [reply grid_map_visualization/src/visualizations/FlatPointCloudVisualization.cpp:46]: Sure will change to false. For future reference, the logic is not going to be affected as the return of the `visualize` method is not checked (see below), but false makes more sense in this context.
https://github.com/ANYbotics/grid_map/blob/e3e00ff932bf8c6a30e1ed5a9e68cb0d5850d56c/grid_map_visualization/src/GridMapVisualization.cpp#L165-L167

File: grid_map_visualization/src/GridMapVisualization.cpp (hunk 4 above)
Comment (Marwan99) [reply grid_map_visualization/src/GridMapVisualization.cpp:126]: Will reverted this change, removed while debugging and forgot to revert it back, my bad.

File: grid_map_visualization/src/GridMapVisualization.cpp (hunk 5 above)
Comment (Marwan99) [reply grid_map_visualization/src/GridMapVisualization.cpp:170]: Will reverted this change, removed while debugging and forgot to revert it back, my bad.

File: grid_map_visualization/include/grid_map_visualization/visualizations/FlatPointCloudVisualization.hpp [hunk 10]
```diff
@@ -6,29 +6,33 @@
...
-#pragma once
+#ifndef GRID_MAP_VISUALIZATION__VISUALIZATIONS__FLATPOINTCLOUDVISUALIZATION_HPP_
+#define GRID_MAP_VISUALIZATION__VISUALIZATIONS__FLATPOINTCLOUDVISUALIZATION_HPP_
 
-#include <grid_map_visualization/visualizations/VisualizationBase.hpp>
 #include <grid_map_core/GridMap.hpp>
 
 // ROS
-#include <ros/ros.h>
+// #include <rclcpp/rclcpp.hpp>
```
[the remark this answers is not in the document]
Comment (Marwan99) [reply grid_map_visualization/include/grid_map_visualization/visualizations/FlatPointCloudVisualization.hpp]: This is a bit trivial suggestion but should we keep just for including what we are using (linter usually asks for that for std libs but doesn't catch it for non-std libs).

File: grid_map_visualization/package.xml [hunk 11]
```diff
@@ -1,21 +1,32 @@
...
+
   <depend>grid_map_core</depend>
-  <depend>grid_map_ros</depend>
   <depend>grid_map_msgs</depend>
-  <depend>tf</depend>
-  <depend>visualization_msgs</depend>
-  <depend>sensor_msgs</depend>
+  <depend>grid_map_ros</depend>
   <depend>nav_msgs</depend>
+  <depend>sensor_msgs</depend>
```
[the remark this answers is not in the document]
Comment (Marwan99) [reply grid_map_visualization/package.xml:22]: Done.

File: grid_map_visualization/src/GridMapVisualization.cpp (hunk 3 above)
Comment (Marwan99) [reply grid_map_visualization/src/GridMapVisualization.cpp:12]: Done.

File: grid_map_visualization/include/grid_map_visualization/GridMapVisualization.hpp [hunk 12]
```diff
@@ -47,10 +49,15 @@ class GridMapVisualization
...
    */
-  void callback(const grid_map_msgs::GridMap& message);
+  void callback(const grid_map_msgs::msg::GridMap::SharedPtr message);
 
- private:
+  /*!
+   * Gets pointer to rclcpp::Node.
+   * @return shared pointer to rclcpp::Node object.
+   */
+  rclcpp::Node::SharedPtr get_node_ptr();
```
[the remark this answers is not in the document]
Comment (Marwan99) [reply grid_map_visualization/include/grid_map_visualization/GridMapVisualization.hpp:58]: I just needed to get a pointer to spin the node, not sure if its common practice though.
https://github.com/ANYbotics/grid_map/blob/c2a4dd3b52a17fc0c03d74e5bbe1117525774d82/grid_map_visualization/src/grid_map_visualization_node.cpp#L17-L19

**Review by maximilianwulf — APPROVED** [review]
Thanks, LGTM.

**Comment by maximilianwulf** [comment]
Thanks, do you know why the pipeline is failing?

**Comment by Marwan99** [comment]
rosbag2_cpp is failing to build, the error seems to be because the container is running out of memory is similar to what we were having before with grid_map_filters. I am testing fixes in a container locally atm.

**Comment by Marwan99** [comment]
@SteveMacenski ready to merge.

File: grid_map_visualization/src/GridMapVisualization.cpp [hunk 13]
```diff
@@ -6,165 +6,170 @@
...
-    : nodeHandle_(nodeHandle),
-      visualizationsParameter_(parameterName),
-      factory_(nodeHandle_),
-      isSubscribed_(false)
+GridMapVisualization::GridMapVisualization(const std::string & parameterName)
+: visualizationsParameter_(parameterName)
+  // isSubscribed_(false)
 {
-  ROS_INFO("Grid map visualization node started.");
+  node_ = std::make_shared<rclcpp::Node>("grid_map_visualization");
```
Comment (SteveMacenski) [inline grid_map_visualization/src/GridMapVisualization.cpp:27]: Why not make the grid map visualization class derive from node? That's pretty typical structure

File: grid_map_visualization/src/GridMapVisualization.cpp [hunk 14]
```diff
@@ -6,165 +6,170 @@
...
+    std::bind(&GridMapVisualization::updateSubscriptionCallback, this));
   initialize();
 }
 
 GridMapVisualization::~GridMapVisualization()
 {
 }
 
+rclcpp::Node::SharedPtr
+GridMapVisualization::get_node_ptr()
```
Comment (SteveMacenski) [inline grid_map_visualization/src/GridMapVisualization.cpp:44]: Why not just let it access `node_`?

File: grid_map_visualization/src/grid_map_visualization_node.cpp [hunk 15]
```diff
@@ -6,17 +6,17 @@
...
 
-  grid_map_visualization::GridMapVisualization gridMapVisualization(nodeHandle, "grid_map_visualizations");
-
-  ros::spin();
+int main(int argc, char ** argv)
+{
+  rclcpp::init(argc, argv);
+  auto node = std::make_shared<grid_map_visualization::GridMapVisualization>(
+    "grid_map_visualizations");
+  rclcpp::spin(node->get_node_ptr());
```
Comment (SteveMacenski) [inline grid_map_visualization/src/grid_map_visualization_node.cpp:19]: If you derive from node, you can just spin `node->get_base_node_interface()`

**Review by SteveMacenski — CHANGES_REQUESTED** [review]
Just some node cleanup and then looks good to merge

File: grid_map_visualization/src/GridMapVisualization.cpp (hunk 13 above)
Comment (Marwan99) [reply grid_map_visualization/src/GridMapVisualization.cpp:27]: The visualisation classes (i.e [FlatPointCloudVisualization](https://github.com/Marwan99/grid_map/blob/visualization-port/grid_map_visualization/src/visualizations/FlatPointCloudVisualization.cpp) ..etc) needed access to the node pointer so they can declare/read parameters, log, check number of subscribers and create publishers. 

I was initially going down the route of making the grid map visualization class derive from node but I was unable to get a pointer to the node created to pass it on to the [visualisations classes](https://github.com/Marwan99/grid_map/tree/visualization-port/grid_map_visualization/src/visualizations), and the only solution I could find was to pass
 `rclcpp::node_interfaces::NodeLoggingInterface`, `rclcpp::node_interfaces::NodeGraphInterface`, `rclcpp::node_interfaces::NodeParametersInterface` and `rclcpp::node_interfaces::NodeBaseInterface` to each [visualization class](https://github.com/Marwan99/grid_map/tree/visualization-port/grid_map_visualization/src/visualizations) to be able to perform the things mentioned above. So instead I went with creating the node object inside the class to be be able to pass it around.

What do you think?

Comment (SteveMacenski) [reply grid_map_visualization/src/GridMapVisualization.cpp:27]: https://github.com/Marwan99/grid_map/blob/visualization-port/grid_map_visualization/include/grid_map_visualization/visualizations/MapRegionVisualization.hpp#L29 they derive from the base class, if that's a node, then each of the visualization is then itself a node.

**Comment by Marwan99** [comment]
@SteveMacenski please have a look.

**Review by SteveMacenski — APPROVED** [review]

**Pull request merged**